### PR TITLE
Update tree example to be hashmap in order to promote immutability

### LINKED ISSUE #1399: Update tree example to be immutable
The Current tree example uses `forceUpdate()` since the nodes are not immutable and are updated without creating a new node for the updated node. This task is to remove `forceUpdate()` and the alternative `_.deepClone`. `deepClone` is the giant happen to make things immutable, but I'd like to promote best practices in your example and plan to use [Normalizr](https://github.com/paularmstrong/normalizr/blob/master/docs/api.md#schema) (e.g.- "hash table") with object spread to inject the one node that needs to be changed with cloning the whole nested tree. 

On a related note, every tree nodes gets rendered due to `shortId.generate()` being called on every render of every node.

Investigation into using a normalized or flattened tree (array of nodes with ids of children) led to finding a bug that prevents `getNodes` from being used correctly

### PR DESCRIPTION
**Pull request by interactivellama**
Fixes #1399

### Additional description
Creates a best practices example for the doc site using a flattened tree (hashmap) that enables better immutability in Flux/Redux state management of the Tree component. Backwards compatible.

[Nested objects shouldn't be used in Redux](https://medium.com/farmdrop/using-normalizr-js-in-a-redux-store-96ab33991369). `deepClone` is an option, but causes a re-render of the whole tree.

The answer appears to be to use `normalizr` to flatten the tree into an array. Then use `getNodes` to walk the React render tree and get the data from the flattened array. This makes the array immutable, but keeps the object references that did not change intact which should prevent React from re-rendering.

* Adds a `log` utility that will route logs to the correct environment such as Storybook actions, `console.log` (on the docs site), or a custom function that we could extend in the future
* Tree: Add additional prop types
* Update Tree sample nodes to be objects. This allows normalizr/getNodes to work correctly
* Add Storybook actions
* Clean up Tree comments
* Tree: Update tests
* Adds more precise tests to getNodes
* Remove warning by adding heading
* Add test to check number of times getNodes to called (to prevent not calling getNodes in the future)
* Update Tree proptypes: allow ids to be numbers and headings to nodes
* Remove Tree as prototype
* Tree: Add normalizr to flatten tree example
* Clarify internal callback args by using options object

---

### Pull Request Review checklist (do not remove)

* [ ] `npm run lint:fix` has been run and linting passes.
* [ ] Mocha, Jest (Storyshots), and `components/component-docs.json` CI tests pass (`npm test`).
* [ ] Tests have been added for new props to prevent regressions in the future. See [readme](https://github.com/salesforce/design-system-react/blob/master/tests/README.md).
* [ ] Review the appropriate Storybook stories. Open [http://localhost:9001/](http://localhost:9001/).
* [ ] Review tests are passing in the browser. Open [http://localhost:8001/](http://localhost:8001/).
* [ ] Review markup conforms to [SLDS](https://www.lightningdesignsystem.com/) by looking at [DOM snapshot strings](https://facebook.github.io/jest/docs/en/snapshot-testing.html).
* [ ] Add year-first date and commit SHA to `last-slds-markup-review` in `package.json` and push.
* [ ] Request a review of the deployed Heroku app by the Salesforce UX Accessibility Team.
* [ ] Add year-first review date, and commit SHA, `last-accessibility-review`, to `package.json` and push.
* [ ] While the contributor's branch is checked out, run `npm run local-update` within locally cloned [site repo](https://github.com/salesforce-ux/design-system-react-site) to confirm the site will function correctly at the next release.

### CODE REVIEW

File: components/index.js [hunk 1]
```diff
@@ -231,3 +231,6 @@ export UtilityIcon from './utilities/utility-icon';
 
 export SLDSTree from './tree';
 export Tree from './tree';
+
+// Library utilities
+export log from '../utilities/log';
```
Comment (garygong) [inline components/index.js]: what is this export used for?

File: components/tree/__docs__/storybook-stories.jsx [hunk 2]
```diff
@@ -1,164 +1,69 @@
...
-					onClick={this.handleClick}
-					onScroll={this.handleScroll}
-					searchTerm={this.state.searchTerm}
-					{...this.props}
-				/>
-			</div>
-		);
-	},
-});
+// View the generated hash map
```
Comment (garygong) [inline components/tree/__docs__/storybook-stories.jsx]: should we clean up the commented `console.logs`?

File: components/index.js (hunk 1 above)
Comment (interactivellama) [reply components/index.js]: Adds a `log` utility that will route logs to the correct environment such as Storybook actions, `console.log` (on the docs site), or a custom function that we could extend in the future. 

My mistake that was missing from the description. This should enable us to use the same stories in Storybook as in the doc website and place the complexity of the multi-environment output within this public function.

File: components/tree/__docs__/storybook-stories.jsx (hunk 2 above)
Comment (interactivellama) [reply components/tree/__docs__/storybook-stories.jsx]: I'd like to keep it, so folks can debug `normalizr`'s output more easily.

**Pull request merged**